### PR TITLE
Bid reject accept beckend | Adniya 

### PR DESCRIPTION
**Pull request by adniyaYousaf**

### CODE REVIEW

File: server/api.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable indent */
```
Comment (soonick) [inline server/api.js]: Remove

File: package.json [hunk 2]
```diff
@@ -51,6 +51,7 @@
 		"db-migrate": "^0.11.14",
 		"del-cli": "^5.1.0",
 		"eslint": "^8.50.0",
+		"eslint-config-prettier": "^9.1.0",
```
Comment (soonick) [inline package.json]: After this diff is merged, please mention in `#launch-module-web-wizards` that the team will need to re-run `npm install`

File: server/api.js [hunk 3]
```diff
@@ -134,4 +135,64 @@ router.get("/tenders", async (req, res) => {
...
+	const itemsPerPage = 10;
+	const totalBids = await db.query(
+		"SELECT COUNT(tender_id) FROM bid WHERE tender_id = $1",
+		[tenderID]
+	);
+	const totalPages = Math.ceil(totalBids.rows[0].count / itemsPerPage);
+	const offset = (page - 1) * itemsPerPage;
+
+	const result = await db.query(
+		"SELECT * FROM bid WHERE tender_id = $1 LIMIT $2 OFFSET $3",
```
Comment (soonick) [inline server/api.js:150]: I'm confused. Are you retrieving tenders or bids? If you are retrieving bids, then the endpoint should be `/bid` and then you can have a `tender_id` query parameter if you want to filter by tender

File: server/api.js [hunk 4]
```diff
@@ -134,4 +135,64 @@ router.get("/tenders", async (req, res) => {
...
+				pagination: {
+					itemsPerPage: 10,
+					currentPage: page,
+					totalPages: totalPages,
+				},
+		  })
+		: res.status(500).send({ code: "SERVER_ERROR" });
+});
+
+router.post("/bid-status/:bidId/:tenderId", async (req, res) => {
```
Comment (soonick) [inline server/api.js]: Change the endpoint to `/bid/:bidId/status`. You don't need the tender id, you can get it based on the bid id

File: server/api.js [hunk 5]
```diff
@@ -134,4 +135,64 @@ router.get("/tenders", async (req, res) => {
...
+	const bidId = parseInt(req.params.bidId);
+	const tenderId = parseInt(req.params.tenderId);
+	let status = req.body.status;
+
+	if (status === "Accepted") {
```
Comment (soonick) [inline server/api.js]: JS doesn't have enums, but we should at least create a constant with all the possible values for `status`, this way we have a centralized place that tells us what is valid. Something like:

```
const valid_statuses = ["Accepted", "Rejected", ...]
```

Then, when you need to validate that the received status is one of those

File: server/api.js [hunk 6]
```diff
@@ -134,4 +135,64 @@ router.get("/tenders", async (req, res) => {
...
+
+		const result = await db.query(
+			"UPDATE bid SET status = $1 WHERE tender_id = $2 AND bid_id = $3;",
+			[status, tenderId, bidId]
+		);
+
+		const rejectedBids = await db.query(
+			"UPDATE bid SET status = $1 WHERE tender_id = $2 AND bid_id != $3;",
+			[reject, tenderId, bidId]
+		);
```
Comment (soonick) [inline server/api.js]: This needs to be done in a transaction

File: server/api.js [hunk 7]
```diff
@@ -134,4 +135,64 @@ router.get("/tenders", async (req, res) => {
...
+
+		result && rejectedBids
+			? res.send({ code: "Success" })
```
Comment (soonick) [inline server/api.js]: Is this a `200`? The success code is unnecessary, the `200` already means it was a success

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js (hunk 1 above)
Comment (adniyaYousaf) [reply server/api.js:1]: Done

File: server/api.js (hunk 3 above)
Comment (adniyaYousaf) [reply server/api.js:150]: Yes! You are right. I have fixed it now.

Comment (adniyaYousaf) [reply server/api.js:150]: Shall I remove the tender_id?

**Comment by soonick** [comment]
@adniyaYousaf , please address all comments before pushing to the branch

File: server/api.js [hunk 8]
```diff
@@ -134,4 +134,97 @@ router.get("/tenders", async (req, res) => {
 	}
 });
 
+router.get("/bid/:tenderId", async (req, res) => {
```
Comment (soonick) [inline server/api.js]: `/bid/1` should refer to the bid with id 1

`/bid?tender_id=1` refers to bids associated with the tender id 1

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js [hunk 9]
```diff
@@ -134,4 +134,97 @@ router.get("/tenders", async (req, res) => {
...
+	const bidId = parseInt(req.params.bidId, 10);
+	const status = req.body.status;
+	const validStatuses = ["Awarded", "Rejected", "Withdraw", "In review"];
+
+	if (!validStatuses.includes(status)) {
+		return res.status(400).send({ code: "INVALID_STATUS" });
+	}
+
+	try {
+		await db.query("BEGIN");
```
Comment (soonick) [inline server/api.js]: The documentation recommends using `pool.connect()` . Let's do the same to be safe. You can look at hos Aisha did it here: https://github.com/LauraSantiag0/Love-Me-Tender/pull/44/files

File: server/api.js [hunk 10]
```diff
@@ -134,4 +134,97 @@ router.get("/tenders", async (req, res) => {
...
+				"UPDATE bid SET status = $1 WHERE tender_id = $2 AND bid_id = $3;",
+				[status, tenderId, bidId]
+			);
+
+			const rejectOtherBidsResult = await db.query(
+				"UPDATE bid SET status = $1 WHERE tender_id = $2 AND bid_id != $3;",
+				[rejectStatus, tenderId, bidId]
+			);
+
+			if (awardBidResult.rowCount > 0 && rejectOtherBidsResult.rowCount > 0) {
```
Comment (soonick) [inline server/api.js]: This should be `||`

**Review by soonick — CHANGES_REQUESTED** [review]

**Review by soonick — APPROVED** [review]

**Pull request merged**